File: Algorithms/Dynamic_Programming/Count_Square_Matrices.py
n = 3
m = 3


def countSquareMatrices(a, N, M):

    count = 0

    for i in range(1, N):
        for j in range(1, M):
            if (a[i][j] == 0):
                continue

            # Calculating  number of square submatrices ending at (i, j)
            a[i][j] = min([a[i - 1][j], a[i][j - 1], a[i - 1][j - 1]])+1

    # Calculate the sum of the array
    for i in range(N):
        for j in range(M):
            count += a[i][j]

    return count


arr = [ [ 1, 0, 1],
    [ 1, 1, 0 ],
    [ 1, 1, 0 ] ]

print(countSquareMatrices(arr, n, m))

File: Algorithms/DP/Count_Square_Matrices.cpp
#include <bits/stdc++.h>
using namespace std;

int main()
{
	//input n: rows and m: columns
	int n, m;
	cin >> n >> m;

	//create array of size n*m and take input from the user
	int ar[n][m];
	for (int i = 0; i < n; i++)
		for (int j = 0; j < m; j++)
			cin >> ar[i][j];

	//initialize count with 0 which keeps count of all submatrices having 1.
	int count = 0;


	for (int i = 1; i < n; i++)
	{
		for (int j = 1; j < m; j++)
		{
			//if ar[i][j] = 0, continue;
			if (ar[i][j] == 0)
				continue;
			//calculate submtrices that end with index (i , j);
			//basically we look at the previous row, column and diagonal and then add 1 to it's minimum.
			else
				ar[i][j] = min({ar[i - 1][j], ar[i][j - 1], ar[i - 1][j - 1]}) + 1;
		}
	}

	//sum of the whole array to get final answer
	for (int i = 0; i < n; i++)
		for (int j = 0; j < m; j++)
			count += ar[i][j];


	//print number of square submatrices containing all zeroes.
	cout << count << "\n";

	return 0;
}
